File: leetcode/0399-evaluate-division/solution.py
from collections import defaultdict
from typing import List


class Solution:
    def calcEquation(self, equations: List[List[str]], values: List[float], queries: List[List[str]]) -> List[float]:
        graph = defaultdict(list)
        for (a, b), value in zip(equations, values):
            graph[a].append((b, value))
            graph[b].append((a, 1.0 / value))

        def dfs(node: str, target: str, product: float, visited: set) -> float:
            if node not in graph:
                return -1.0
            if node == target:
                return product
            visited.add(node)
            for neighbor, weight in graph[node]:
                if neighbor not in visited:
                    result = dfs(neighbor, target, product * weight, visited)
                    if result != -1.0:
                        return result
            return -1.0

        results = []
        for c, d in queries:
            if c not in graph or d not in graph:
                results.append(-1.0)
            else:
                results.append(dfs(c, d, 1.0, set()))
        return results

File: leetcode/0399-evaluate-division/solution.cpp
#include <vector>
#include <string>
#include <unordered_map>
#include <unordered_set>
using namespace std;

class Solution {
public:
    vector<double> calcEquation(vector<vector<string>>& equations, vector<double>& values, vector<vector<string>>& queries) {
        unordered_map<string, unordered_map<string, double>> graph;

        for (size_t i = 0; i < equations.size(); i++) {
            const string& a = equations[i][0];
            const string& b = equations[i][1];
            double value = values[i];
            graph[a][b] = value;
            graph[b][a] = 1.0 / value;
        }

        vector<double> results;
        for (auto& query : queries) {
            const string& c = query[0];
            const string& d = query[1];
            if (!graph.count(c) || !graph.count(d)) {
                results.push_back(-1.0);
            } else {
                unordered_set<string> visited;
                results.push_back(dfs(graph, c, d, 1.0, visited));
            }
        }
        return results;
    }

private:
    double dfs(unordered_map<string, unordered_map<string, double>>& graph, const string& node,
               const string& target, double product, unordered_set<string>& visited) {
        if (node == target) {
            return product;
        }
        visited.insert(node);
        for (auto& [neighbor, weight] : graph[node]) {
            if (!visited.count(neighbor)) {
                double result = dfs(graph, neighbor, target, product * weight, visited);
                if (result != -1.0) {
                    return result;
                }
            }
        }
        return -1.0;
    }
};
